### ISSUE TITLE
`Factory.getInstance().getComponents` creates singleton component multiple times

### ISSUE THREAD

**Issue by fedotxxl** [restx/restx#236]
1. Mark `@Component` by any interface
2. Add `Factory.getInstance().getComponents(YourClass.class).forEach(this::doSomething)` to `start` of `AutoStartable` component
3. Start project in `prod` mode

Expectations:
Service should be created once

Result:
`Factory.getInstance().getComponents` will create new instance of service

Sample application:
https://github.com/fedotxxl/restx-bugs

Start log:

```
18:28:05.718 [main] [] INFO  o.e.j.server.handler.ContextHandler - started o.e.j.w.WebAppContext{/,file:/home/fbelov/job/projects-test/restx-bugs/src/main/webapp/},src/main/webapp
18:28:05.718 [main] [] INFO  o.e.j.server.handler.ContextHandler - started o.e.j.w.WebAppContext{/,file:/home/fbelov/job/projects-test/restx-bugs/src/main/webapp/},src/main/webapp
18:28:06.174 [main] [] WARN  io.belov.restx.SampleService - Creating SampleService - io.belov.restx.SampleService@2b9ed6da
18:28:06.934 [main] [] INFO  o.h.validator.internal.util.Version - HV000001: Hibernate Validator 5.0.1.Final
18:28:06.949 [main] [] INFO  o.h.v.i.engine.ConfigurationImpl - HV000002: Ignoring XML configuration.
18:28:07.043 [main] [] INFO  io.belov.restx.Bootstrap - Bootstrap starting
18:28:07.052 [main] [] WARN  io.belov.restx.SampleService - Creating SampleService - io.belov.restx.SampleService@64c2b546
18:28:07.145 [main] [] INFO  io.belov.restx.Bootstrap - Bootstrap finished
18:28:07.210 [main] [] INFO  o.e.jetty.server.AbstractConnector - Started SelectChannelConnector@0.0.0.0:8091
```

As you can see `SampleService` was created twice.

To start application:
1. `git clone`
2. `gradle idea`
3. Application configuration with main class `io.belov.restx.AppServer` and vm options -`Drestx.mode=prod`

**Comment by fcamblor** [comment]
(Sorry for the long delay)

Yes that's a known limitation of `AutoStartable`s

`AutoStartable`s are intended to be loaded in separate classloader (dedicated to component not having to be dropped away, [see doc](http://restx.io/docs/ref-lifecycle.html#toc8))

Also, downside for `AutoStartable`s is : if you inject some dependencies into it (either by constructor, or, as you do, with a manual `Factory.getInstance().getComponents()` call), then those dependencies will rely on a `Factory` attached to current classloader (the one attached to the `AutoStartable`, which is specific)

This explains why constructor of your `SampleService` has been called twice :
- It has been called the first time with the `AutoStartable`'s classloader
- It has then been called when injection into `SampleResource` is requested, using a non-`AutoStartable` classloader


tldr; `AutoStartable` should be standalone components as much as possible (when I say "standalone", it means that no `@Component` dependencies should be injected into it)


If you need something which should be called only once, during application startup, and which should have access to your whole business logic, then I would suggest you to :
- create a dedicated REST endpoint doing this startup logic
- this REST endpoint should be annotated with `@PermitAll` for convenience, but should take a secret token as query parameter
- you should then call this endpoint from within your `AutoStartable`, with the secret token, using a simple HTTP client lib (like `okhttp`, `kevinsawicki's http-request` or simply `commons-http`), and after a certain delay in order to be sure that your app has finished its startup (see ScheduledExecutorService.schedule() for this purpose)
